text/3700-ptr-assume-moved.md

```diff
@@ -0,0 +1,257 @@
+> any other object accessible to the caller.
+
+This will result in the new pointer getting a brand new provenance, disjoint from the provenance of
+the original pointer.
+
+Every change to the high bits has to at least simulate a move operation and we must ensure the old
+pointers are invalidated. This is due to the aforementioned discrepancy between how Rust & LLVM see
+a memory address and how the OS & hardware see memory addresses.  From the OS & hardware
+perspective, the high bits are reserved for metadata and do not actually form part of the address
+(in the sense of an 'address' being an index into the memory array).  From the LLVM perspective, the
+high bits are part of the address and changing them means we are now dealing with a different
+address altogether.  Having to reconcile those two views necessarily creates some friction and extra
+considerations.
+
+Function signature, documentation and implementation:
+
+```
+/// Assume that the object pointed to by a pointer has been moved to a new address
+///
+/// Intended for use with pointer tagging architecture features such as AArch64 TBI.
+/// This function creates a new pointer with the address `new_address` and a brand new provenance,
+/// then assumes that a move from the original address to the new address has taken place.
+/// Note that this is only an indication for the compiler - nothing actually gets moved or reallocated.
+///
+/// SAFETY: Users *must* ensure that `new_address` actually contains the same memory as the original.
+/// The primary use-case is working with various architecture pointer tagging schemes, where two
+/// different 64-bit addresses can point to the same chunk of memory due to some bits being ignored.
+/// When used incorrectly, this function can be used to violate the memory model in arbitrary ways.
+/// Furthermore, after using this function, users must ensure that the underlying memory is only ever
+/// accessed through the newly created pointer. Any accesses through the original pointer
+/// (or any pointers derived from it) would be Undefined Behaviour.
+/// Additionally, this function is only designed for use with heap allocations. Trying to use it with
+/// a pointer to a stack-allocated variable will almost certainly result in Undefined Behaviour and
+/// should not be done under any circumstances.
+#[inline(never)]
+#[unstable(feature = "ptr_assume_moved", issue = "none")]
+#[cfg_attr(not(bootstrap), rustc_simulate_allocator)]
+#[allow(fuzzy_provenance_casts)]
+pub unsafe fn assume_moved<T>(original: *mut T, new_address: usize) -> *mut T {
+    // FIXME(strict_provenance_magic): I am magic and should be a compiler intrinsic.
+    let mut ptr = new_address as *mut T;
+    // SAFETY: This does not do anything
+    unsafe {
+        asm!(
+            "/* simulate a move from {original} to {ptr} */",
+            original = in(reg) original,
+            ptr = inout(reg) ptr
+        );
+    }
```

Comment:
Is this equivalent to just calling `black_box`?

Give that `black_box` is documented as just a hint, this should be marked as a hack that we use for codegen backends which don't have a proper way to represent this operation.

We probably want to make this an intrinsic, and then each backend can decide to implement this either like black_box or in a more direct way. The RFC shouldn't specify the exact implementation, but instead describe the concerns backends have to be aware of here.

Reply:
I'm not sure whether it's 1:1 equivalent to black_box, this was originally suggested by @Amanieu as a solution to the problem. I'll update the RFC with a note about this being just an example implementation and the concerns as requested.

Reply:
i expect it to not be equivalent to `black_box` (since that's documented to be just a hint), but to be equivalent to the inline asm that `black_box` is currently implemented with since rustc and llvm theoretically have no way of knowing you aren't calling `realloc` from inside the inline asm.

Reply:
Yes that's what I meant, I should have been more precise -- a backend can decide that its black_box impl is sufficient for this purpose, but a user cannot assume them to be equivalent.